docs/CppInteroperability/SwiftToCppInteroperabilityVisionDocument.md

```diff
@@ -0,0 +1,137 @@
+
+Swift's support for C++ interoperability is modeled after its support for C and Objective-C interoperability. For "forward" interoperability, the embedded Clang compiler is used to directly load C++ Clang modules and translate declarations into a native Swift representation. The "forward" interoperability [vision document](https://github.com/apple/swift/pull/60501/files) provides more details about this model. For "reverse" interoperability, Swift generates a header that uses C++ language constructs to represent Swift APIs that are exposed by the Swift module. Because C++ is much more expressive and flexible than Objective-C, the generated header is able to provide representation for native Swift functions, methods, initializers, accessors and types. This allows C++ programmers to call into Swift APIs using the familiar C++ function and member function call syntax.
+
+### C++ interoperability is opt-out not opt-in
+
+The user model employed by Swift-to-C++ interoperability is different than the user model employed by the Objective-C interoperability. C++ can use Swift's native calling convention when calling Swift functions, and C++ does not have restrictions on what Swift types it can work with. This means that Swift APIs do not require additional annotations when they're bridged to C++. This fact highlights the key difference between C++ and Objective-C interoperability; Swift library author is not required to change their code when their APIs are exposed to C++, whereas they are required to opt-in into interoperability using an `@objc` annotation when exposing APIs to Objective-C. Swift library authors will still be able to opt-out from exposing certain APIs to C++ when they desire to do so.
+
+The permissive user model of Swift-to-C++ interoperability implies that any Swift library that can be imported into Swift can also be "imported" into C++, as long as its APIs are representable by the C++ code in the generated header. The Swift compiler can generate this header from the library's module interface file upon user's request, even when the library's author did not have such use case in mind.
+
+## Goals
+
+Swift and C++ are both feature rich programming languages. They share many similar features, but Swift has some features that lack close analogues in C++. The primary goal of Swift-to-C++ interoperability is to expose every language feature of Swift for which a reasonable C++ API can be created to represent that feature. For example, C++ does not have a concept of argument labels, but the argument labels can be integrated into the base name of a function to get a roughly similar effect. On the other hand, C++ does not have any feature that can represent a Swift feature like result builders in a reasonable manner. Therefore, APIs that use Swift features like result builders must be dropped in the C++ interface to the module.
```

Comment:
Do we actually need to drop something like result builders? That seems to me like a feature that we could support on the C++ side in the "desugared" form, which we already permit people to write even in Swift. I.e., you wouldn't get the nice declarative syntax (or any of the other helpful glue that result builders pull together for you), but you could write out the equivalent post-transform code yourself if you wanted. Have I misunderstood this paragraph?

Reply:
Sure, that's a valid argument. Fundamentally result builders can be exposed that way. I think it might be better to make a stronger case for another language feature. I updated the doc to talk about macros instead.